exporter/gcloudpubsubexporter/exporter.go

```diff
@@ -0,0 +1,166 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcloudpubsubexporter
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
```

Comment:
Shopify has been running a Google PubSub (trace) exporter in production for almost a year. The largest improvement in stability came from switching to the lower level package `cloud.google.com/go/pubsub/apiv1`. The high level package used here provides an asynchronous `Publish` method that does not allow back pressure in the collector, and limits the usefulness of the `memorylimiter` processor, queued retry, etc. The lower level package provides a synchronous `Publish` call, which is more appropriate for the collector.

cc @ibawt

Reply:
Thanks, @fbogsany I will certainly rewrite this as I was a bit struggling with this. I'm writing the OTEP now and will refer to this in this thread.

Reply:
> Shopify has been running a Google PubSub (trace) exporter in production for almost a year. The largest improvement in stability came from switching to the lower level package `cloud.google.com/go/pubsub/apiv1`. The high level package used here provides an asynchronous `Publish` method that does not allow back pressure in the collector, and limits the usefulness of the `memorylimiter` processor, queued retry, etc. The lower level package provides a synchronous `Publish` call, which is more appropriate for the collector.
> 
> cc @ibawt

At Shopify do you push the complete OTLP Request over PubSub or are you unraffling the content. Let me elaborate: For spans, the `ExportTraceServiceRequest` contains a repeated `ResourceSpans`. The request is already a mini-batch. I'm going to propose to put the `ResourceSpans` on Pubsub (or any message bus).

Reply:
We publish a full `ExportTraceServiceRequest`. We treat PubSub as a transport for OTLP, which makes things easier for downstream consumers. In particular, we can reuse code between the collector and the consumer. For example, we have an exporter from PubSub to GCS (Avro) that we'll eventually move into the collector, so it helps if that receives the data it would receive if it were in the collector instead.

Specifically, for `traces pdata.Traces`, we `Publish` the gzipped bytes returned from `traces.ToOtlpProtoBytes()`.

Reply:
While it's a big effort, I do wish that OTLP had a `FullSpan` or similar type which was a normalized span with resource / instrumentation library built in. Having the ability to pack a self-contained span can help with cases like message queues, downstream processing like metrics updating, or even updating a trace with a new span, generally model better with span-per-message when dealing with retries and such.